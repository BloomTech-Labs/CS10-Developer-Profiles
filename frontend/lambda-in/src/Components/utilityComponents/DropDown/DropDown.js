import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import Icon from '@material-ui/core/Icon';

/**
 * A wrapper component for @material-ui <ExpansionPanel />
 *
 * @description A Dropdown component wrapper, which accepts any type of children.
 *
 * @function A functional component.
 * @prop {string} header - Text to be displayed when the Dropdown is collapsed.
 *
 * @example How to use it
 * <DropDown header="Header text to display here">
 */
const DropDown = ({ header, children }) => (
  <Fragment>
    <div className="panelMargin">
      <ExpansionPanel>
        <ExpansionPanelSummary expandIcon={<Icon>add_circle</Icon>}>
          <Typography>{header}</Typography>
        </ExpansionPanelSummary>
        <ExpansionPanelDetails>
          <div className="expansionPanel">{children}</div>
        </ExpansionPanelDetails>
      </ExpansionPanel>
    </div>
  </Fragment>
);

DropDown.propTypes = {
  header: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
    PropTypes.func,
  ]).isRequired,
};

export default DropDown;
