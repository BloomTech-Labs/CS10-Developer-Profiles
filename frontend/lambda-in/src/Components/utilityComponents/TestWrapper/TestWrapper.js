import React, { Component, Fragment } from 'react';

/**
 * A Functional stateless component wrapper
 *
 * @description Enzyme's 'shallow' method don't create an instance for
 * stateless functional componets
 * Whitout an instance is not possible to access the functional componet's
 * props.
 * Because enzyme's shallow generate an instance for class coponents, <TestWrapper/>
 * allow us to access any functional coponent passed as a children.
 * @see https://github.com/airbnb/enzyme/issues/1232#issuecomment-334510754
 */
class TestWrapper extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  /**
   * Catch children component error an log them in test-enviroment
   *
   * @description An Error bundary to catch potential children component's
   * Javascrip errors
   * @see https://reactjs.org/docs/error-boundaries.html
   * @param {object} error - Error ebject
   * @param {object} info - Component stack on error
   */
  componentDidCatch(error, info) {
    // eslint-disable-next-line no-console
    console.log('Error Bundary', error, info);
  }

  render() {
    // eslint-disable-next-line react/prop-types
    const { children } = this.props;
    return <Fragment>{children({ ...this.props })}</Fragment>;
  }
}

export default TestWrapper;
