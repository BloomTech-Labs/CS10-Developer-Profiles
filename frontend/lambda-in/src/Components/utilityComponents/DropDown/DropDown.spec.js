import React from 'react';
import ReactTestUtils from 'react-dom/test-utils';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import DropDown from './DropDown';
import TestWrapper from '../TestWrapper/TestWrapper';

configure({ adapter: new Adapter() });

/**
 * @author jesuarva
 */
describe('DropDown component', () => {
  const dropDown = <DropDown header="Testing DropDown component" />;

  it('should renders correctly', () => {
    const tree = renderer.create(dropDown).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should have a `header` props', () => {
    const instance = ReactTestUtils.renderIntoDocument(
      <TestWrapper>{() => dropDown}</TestWrapper>,
    );
    const p = ReactTestUtils.findRenderedDOMComponentWithTag(instance, 'p');

    expect(p.innerHTML).toMatch('Testing DropDown component');
  });
});
