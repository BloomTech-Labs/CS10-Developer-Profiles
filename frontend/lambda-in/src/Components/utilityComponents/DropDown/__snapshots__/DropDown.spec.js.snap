// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown component should renders correctly 1`] = `
<div
  className="panelMargin"
>
  <div
    className="MuiPaper-root-4 MuiPaper-elevation1-7 MuiExpansionPanel-root-1"
  >
    <div
      aria-expanded={false}
      className="MuiButtonBase-root-37 MuiExpansionPanelSummary-root-31"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex="0"
    >
      <div
        className="MuiExpansionPanelSummary-content-35"
      >
        <p
          className="MuiTypography-root-40 MuiTypography-body1-49"
        >
          Testing DropDown component
        </p>
      </div>
      <div
        aria-hidden="true"
        className="MuiButtonBase-root-37 MuiIconButton-root-66 MuiExpansionPanelSummary-expandIcon-36"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={-1}
      >
        <span
          className="MuiIconButton-label-71"
        >
          <span
            aria-hidden="true"
            className="material-icons MuiIcon-root-72"
          >
            add_circle
          </span>
        </span>
        <span
          className="MuiTouchRipple-root-81"
        />
      </div>
    </div>
    <div
      aria-hidden="true"
      className="MuiCollapse-container-88"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper-90"
      >
        <div
          className="MuiCollapse-wrapperInner-91"
        >
          <div
            className="MuiExpansionPanelDetails-root-92"
          >
            <div
              className="expansionPanel"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
